### DISCUSSION TITLE
Event Listeners not being triggered after adding API

### DISCUSSION THREAD

**Discussion by irishmango** [community/community#102987] [Apps, API and Webhooks]
### Select Topic Area

Bug

### Body

I'm building a simple store website where I need the euro price converted into bitcoin when the total is shown in the shopping cart. Everything was working until I added the eur-btc API, but now the 'add to cart' and 'show cart' buttons arent working.

Here are my gists for the html and javascript: 
https://gist.github.com/irishmango/d89dbf003429793fafa7a9afac57506d - html https://gist.github.com/irishmango/cf7cd76306070c2b3d63f83660826f7d - js

I thought that maybe the error I was due to the use of await outside of the async function. So I tried to wrap my code in an immediately-invoked async function. Still nothing. As far I can see, all IDs match up.

Co-Pilot also hasnt been able to help fix the problem. 

I've tried everything I can think of, but nothing seems to work.

Anyone have any ideas?

Still kinda new to coding so would appreciate any help(and patience) you have :)

**Comment by amankr1469** [comment]
Move the code that adds event listeners for "Add to Cart" buttons and "Show Cart" button inside the then block of the API call. This ensures that these event listeners are added only after the API response is received.

**Reply by joaomarcosjova** [reply]
Good point also

**Comment by joaomarcosjova** [comment]
Sure! Let's break down the JavaScript code provided in your gist and see how we can improve it. We'll focus on organizing the code, handling API calls, and ensuring proper event binding.

```javascript
// Wrap the entire code in an immediately-invoked function expression (IIFE) to encapsulate scope
(function() {
    // Define variables for DOM elements
    const cartButton = document.getElementById('add-to-cart');
    const showCartButton = document.getElementById('show-cart');

    // Define function to fetch Bitcoin price from API
    async function getBitcoinPrice() {
        try {
            const response = await fetch('https://api.coindesk.com/v1/bpi/currentprice.json');
            const data = await response.json();
            const bitcoinPrice = data.bpi.EUR.rate; // Extract Bitcoin price from response
            return bitcoinPrice;
        } catch (error) {
            console.error('Error fetching Bitcoin price:', error);
            return null;
        }
    }

    // Define function to handle adding to cart
    function addToCart() {
        // Implement functionality to add item to cart
        console.log('Item added to cart');
    }

    // Define function to handle showing cart
    function showCart() {
        // Implement functionality to show cart
        console.log('Showing cart');
    }

    // Add event listener for 'Add to Cart' button
    cartButton.addEventListener('click', addToCart);

    // Add event listener for 'Show Cart' button
    showCartButton.addEventListener('click', showCart);

    // Immediately invoke function to fetch Bitcoin price and update UI
    (async function() {
        const bitcoinPrice = await getBitcoinPrice();
        if (bitcoinPrice) {
            // Update UI to display Bitcoin price
            console.log('Bitcoin price:', bitcoinPrice);
        } else {
            // Handle error or display fallback message
            console.log('Failed to fetch Bitcoin price');
        }
    })();
})();
```

Here's a breakdown of the changes and improvements:

1. **Encapsulation with IIFE:**
   The entire JavaScript code is wrapped within an immediately-invoked function expression (IIFE). This helps encapsulate the scope of variables and functions, preventing them from polluting the global namespace.

2. **Variable Definitions:**
   DOM elements (`cartButton` and `showCartButton`) are cached using `document.getElementById` to improve performance.

3. **Asynchronous API Call:**
   The `getBitcoinPrice` function is defined as an asynchronous function using `async` and `await` to fetch the Bitcoin price from the API. Error handling is implemented using `try...catch`.

4. **Event Listeners:**
   Event listeners for the 'Add to Cart' and 'Show Cart' buttons are added using `addEventListener`.

5. **Fetch Bitcoin Price on Page Load:**
   The function to fetch the Bitcoin price is immediately invoked when the script runs. Upon successful retrieval of the price, the UI can be updated accordingly.

6. **Logging and Error Handling:**
   Console logs are used for debugging purposes, and error messages are logged if there are issues with fetching the Bitcoin price.

By organizing the code in this manner, you can improve readability, maintainability, and reliability of your JavaScript code for the store website. Feel free to adjust the implementation based on your specific requirements and application structure!

**Reply by irishmango** [reply]
Hi there, sorry for the delay in responding to your comment. I appreciate you taking the time to help me!

I tried implementing your code, fixed some errors that popped up. But the problem remains. There is no reaction if I click add to cart or show cart. 
I tried changing those two buttons to classes as apposed to ids. Still nothing. 

Would you be able to advise further? I have updated my Gists:
https://gist.github.com/irishmango/d89dbf003429793fafa7a9afac57506d - html https://gist.github.com/irishmango/cf7cd76306070c2b3d63f83660826f7d - js
